modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java

```diff
@@ -1010,6 +1017,38 @@ private IgniteCheckedException shutdown(PageStore store, boolean cleanFile, Igni
         return aggr;
     }
 
+    /**
+     * Return cache store holedr.
+     *
+     * @param grpId Cache group ID.
+     * @return Cache store holder.
+     */
+    private CacheStoreHolder getHolder(int grpId) {
+        return idxCacheStores.computeIfAbsent((grpId), (key) -> {
+            CacheGroupDescriptor gDesc = cctx.cache().cacheGroupDescriptors().get(grpId);
+
+            CacheStoreHolder holder0 = null;
+
+            if (gDesc != null) {
+                Integer cacheId = F.firstValue(gDesc.caches());
+
+                if (cacheId != null) {
+                    DynamicCacheDescriptor cacheDesc = cctx.cache().cacheDescriptor(cacheId);
+
+                    if (CU.isPersistentCache(cacheDesc.cacheConfiguration(), cctx.gridConfig().getDataStorageConfiguration())) {
+                        try {
+                            holder0 = initForCache(cacheDesc.groupDescriptor(), cacheDesc.cacheConfiguration());
+                        } catch (IgniteCheckedException e) {
+                            throw new IgniteException(e);
```

Comment:
What will happen if exception is thrown here? Previously when we initialized everything beforehand we had a clear path: node fails to restore memory and shuts down.

Now with this change we don't even have a checked exception to catch, it is wrapped into unchecked.
Should we call FailureHandler directly in this case if something goes wrong during lazy initialization of cache store?

Reply:
I  change a bit FilePageStoreManager#initDir cached exceptions, hope it would be enough.